lio_sam:

  # Topics
  pointCloudTopic: "/carla/ego_vehicle/semantic_lidar"               # Point cloud data
  imuTopic: "/carla/ego_vehicle/imu"                         # IMU data

  odomTopic: "/odometry/imu"                   # IMU pre-preintegration odometry, same frequency as IMU

  gpsTopic: "odometry/gpsz"                   # GPS odometry topic from navsat, see module_navsat.launch file

  # Frames
  lidarFrame: "base_link"
  baselinkFrame: "base_link"
  odometryFrame: "odom"
  mapFrame: "liosam_map"

  # GPS Settings
  useImuHeadingInitialization: true           # if using GPS data, set to "true"
  useGpsElevation: false                      # if GPS elevation is bad, set to "false"
  gpsCovThreshold: 2.0                        # m^2, threshold for using GPS data
  poseCovThreshold: 25.0                      # m^2, threshold for using GPS data
  
  # Export settings
  savePCD: false                              # https://github.com/TixiaoShan/LIO-SAM/issues/3
  savePCDDirectory: "/LOAM/"        # in your home folder, starts and ends with "/". Warning: the code deletes "LOAM" folder then recreates it. See "mapOptimization" for implementation

  # Sensor Settings
  sensor: carla                            # lidar sensor type, either 'velodyne' or 'ouster' or 'carla'
  N_SCAN: 64                                  # number of lidar channel (i.e., 16, 32, 64, 128)
  Horizon_SCAN: 1800                          # lidar horizontal resolution (Velodyne:1800, Ouster:512,1024,2048)
  downsampleRate: 2                           # default: 1. Downsample your data if too many points. i.e., 16 = 64 / 4, 16 = 16 / 1
  lidarMinRange: 0.1                          # default: 1.0, minimum lidar range to be used
  lidarMaxRange: 50.0                       # default: 1000.0, maximum lidar range to be used
  minAngle: -26.8
  maxAngle: 2.0
  
  # Vehicle settings
  egoWidth: 2.17
  egoLength: 4.809
  lidarOffsetLength: -0.2 # in x-axis usually
  lidarOffsetWidth: 0.0 # in y-axis usually

  # IMU Settings
  imuAccNoise: 3.9939570888238808e-03
  imuGyrNoise: 1.5636343949698187e-03
  imuAccBiasN: 6.4356659353532566e-05
  imuGyrBiasN: 3.5640318696367613e-05
  imuGravity: 9.80511
  imuRPYWeight: 0.01

  # Extrinsics (lidar -> IMU)
  extrinsicTrans: [-0.3, 0.0, -0.3]
  extrinsicRot: [1, 0, 0, 0, 1, 0, 0, 0, 1]
  extrinsicRPY: [1, 0, 0, 0, 1, 0, 0, 0, 1]

  # Extrinsics (IMU -> GPS)
  imu2gpsTrans: [0.144, -0.161, 0.769]

  # LOAM feature threshold
  edgeThreshold: 1.0
  surfThreshold: 0.1
  edgeFeatureMinValidNum: 10
  surfFeatureMinValidNum: 100

  # voxel filter paprams
  odometrySurfLeafSize: 0.4                     # default: 0.4 - outdoor, 0.2 - indoor
  mappingCornerLeafSize: 0.2                    # default: 0.2 - outdoor, 0.1 - indoor
  mappingSurfLeafSize: 0.4                      # default: 0.4 - outdoor, 0.2 - indoor

  # robot motion constraint (in case you are using a 2D robot)
  z_tollerance: 1000                            # meters
  rotation_tollerance: 1000                     # radians

  # CPU Params
  numberOfCores: 10                              # number of cores for mapping optimization
  mappingProcessInterval: 0.15                  # seconds, regulate mapping frequency

  # Surrounding map
  surroundingkeyframeAddingDistThreshold: 1.0   # meters, regulate keyframe adding threshold
  surroundingkeyframeAddingAngleThreshold: 0.2  # radians, regulate keyframe adding threshold
  surroundingKeyframeDensity: 2.0               # meters, downsample surrounding keyframe poses   
  surroundingKeyframeSearchRadius: 50.0         # meters, within n meters scan-to-map optimization (when loop closure disabled)

  # Loop closure
  loopClosureEnableFlag: true
  loopClosureFrequency: 1.0                     # Hz, regulate loop closure constraint add frequency
  surroundingKeyframeSize: 50                   # submap size (when loop closure enabled)
  historyKeyframeSearchRadius: 15.0             # meters, key frame that is within n meters from current pose will be considerd for loop closure
  historyKeyframeSearchTimeDiff: 30.0           # seconds, key frame that is n seconds older will be considered for loop closure
  historyKeyframeSearchNum: 25                  # number of hostory key frames will be fused into a submap for loop closure
  historyKeyframeFitnessScore: 0.3              # icp threshold, the smaller the better alignment

  # Visualization
  globalMapVisualizationSearchRadius: 1000.0    # meters, global map visualization radius
  globalMapVisualizationPoseDensity: 10.0       # meters, global map visualization keyframe density
  globalMapVisualizationLeafSize: 1.0           # meters, global map visualization cloud density

cluster_debug: false
tracking_debug: false

# ----------------------
# TRACKING
# ----------------------
tracking:
  enableTracking: true
  clustering: 
    method: "graph" # 'euclidean' or 'graph' or 'slr'
    nongroundDownSample: 0.05 # opt: 0.05
    maxZ: 3.0
    minZ: -10.0
    maxR: 100.0
    clusteringTolerance: 1.0 # opt 0.5
    minClusterSize: 5 # opt 5
    maxClusterSize: 5000 # opt 5000
    graphDistThres: 0.2 # horizontal (opt 0.3)
    graphMergeThres: 1.2 # vertical
    graphLookupSize: 6 # opt 6
    graphAngleThres: 20.0
    searchWindowSize: 4 # opt 4

  # GMPHD filter
  filter:
    meanXBirth: 500.0 # opt: 500
    meanYBirth: 500.0 # opt: 500
    varPoseBirth: 2500
    varVelBirth: 2500
    weightBirth: 0.2
    samplingPeriod: 0.1
    processNoise: 1e-12
    spawnPoseNoise: 90 # opt: 90 mapping: 40
    spawnVelNoise: 20 # opt: 20 mapping: 10
    detectionProb: 1.0
    measPoseNoise: 1.0
    measSpeedNoise: 4.0
    pruneThres: 0.2 # opt: 0.2
    pruneMergeThres: 0.0005 # opt: 0.0005
    maxGaussian: 150
    survivalProb: 1.0
    minWeightTrack: 0.0 # opt: 0.0

  # L-Shape fitting
  lshape:
    minHeight: 0.3 # opt: 0.5
    maxHeight: 3.0
    maxArea: 50.0 
    maxRatio: 9.0
    minDensity: 3.0
    maxDensity: 1500

# ----------------------
# REMOVAL
# ----------------------
removal:
  enableDynamicRemoval: true
  velThres: 0.05
  modeProbThres: 0.5

# ----------------------
# EXPERIMENT
# ----------------------
experiment:
  slam_save_dir: "/home/euigon/experiment/traj_result_dynamic/city/"
  trk_save_dir: "/home/euigon/experiment/trk_result/"
  cluster_save_dir: "/home/euigon/experiment/cls_result/"
  eval_tracking: false
  eval_slam: true
  eval_clustering: false
  max_out_cnt: 20
  lidar_scope: 40.0
  min_pt_cnt: 20 # opt 30
  gnd_ratio: 0.9
  

# Navsat (convert GPS coordinates to Cartesian)
navsat:
  frequency: 50
  wait_for_datum: false
  delay: 0.0
  magnetic_declination_radians: 0
  yaw_offset: 0
  zero_altitude: true
  broadcast_utm_transform: false
  broadcast_utm_transform_as_parent_frame: false
  publish_filtered_gps: false

# EKF for Navsat
ekf_gps:
  publish_tf: false
  map_frame: map
  odom_frame: odom
  base_link_frame: base_link
  world_frame: odom

  frequency: 50
  two_d_mode: false
  sensor_timeout: 0.01
  # -------------------------------------
  # External IMU:
  # -------------------------------------
  imu0: imu_correct
  # make sure the input is aligned with ROS REP105. "imu_correct" is manually transformed by myself. EKF can also transform the data using tf between your imu and base_link
  imu0_config: [false, false, false,
                true,  true,  true,
                false, false, false,
                false, false, true,
                true,  true,  true]
  imu0_differential: false
  imu0_queue_size: 50 
  imu0_remove_gravitational_acceleration: true
  # -------------------------------------
  # Odometry (From Navsat):
  # -------------------------------------
  odom0: odometry/gps
  odom0_config: [true,  true,  true,
                 false, false, false,
                 false, false, false,
                 false, false, false,
                 false, false, false]
  odom0_differential: false
  odom0_queue_size: 10

  #                            x     y     z     r     p     y   x_dot  y_dot  z_dot  r_dot p_dot y_dot x_ddot y_ddot z_ddot
  process_noise_covariance: [  1.0,  0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    1.0,  0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    10.0, 0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0.03, 0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0.03, 0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0.1,  0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0.25,  0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0.25,  0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0.04,  0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0.01, 0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0.01, 0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0.5,  0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0.01, 0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0.01,   0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0.015]

rosbag:
  type: "carla"
